File: src/main/java/com/pubnub/api/endpoints/objects_api/spaces/GetSpace.java
package com.pubnub.api.endpoints.objects_api.spaces;

import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubException;
import com.pubnub.api.builder.PubNubErrorBuilder;
import com.pubnub.api.endpoints.Endpoint;
import com.pubnub.api.enums.PNOperationType;
import com.pubnub.api.enums.PNSpaceFields;
import com.pubnub.api.managers.RetrofitManager;
import com.pubnub.api.managers.TelemetryManager;
import com.pubnub.api.managers.token_manager.PNResourceType;
import com.pubnub.api.managers.token_manager.TokenManagerProperties;
import com.pubnub.api.managers.token_manager.TokenManagerPropertyProvider;
import com.pubnub.api.models.consumer.objects_api.space.PNGetSpaceResult;
import com.pubnub.api.models.consumer.objects_api.space.PNSpace;
import com.pubnub.api.models.consumer.objects_api.util.InclusionParamsProvider;
import com.pubnub.api.models.server.objects_api.EntityEnvelope;
import lombok.Setter;
import lombok.experimental.Accessors;
import retrofit2.Call;
import retrofit2.Response;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Accessors(chain = true, fluent = true)
public class GetSpace extends Endpoint<EntityEnvelope<PNSpace>, PNGetSpaceResult>
        implements InclusionParamsProvider<GetSpace, PNSpaceFields>,
        TokenManagerPropertyProvider {

    private Map<String, String> extraParamsMap;

    @Setter
    private String spaceId;

    public GetSpace(PubNub pubnubInstance, TelemetryManager telemetry, RetrofitManager retrofitInstance) {
        super(pubnubInstance, telemetry, retrofitInstance);
        extraParamsMap = new HashMap<>();
    }

    @Override
    protected List<String> getAffectedChannels() {
        return null;
    }

    @Override
    protected List<String> getAffectedChannelGroups() {
        return null;
    }

    @Override
    protected void validateParams() throws PubNubException {
        if (this.getPubnub().getConfiguration().getSubscribeKey() == null
                || this.getPubnub().getConfiguration().getSubscribeKey().isEmpty()) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_SUBSCRIBE_KEY_MISSING).build();
        }

        if (this.spaceId == null || this.spaceId.isEmpty()) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_SPACE_ID_MISSING).build();
        }
    }

    @Override
    protected Call<EntityEnvelope<PNSpace>> doWork(Map<String, String> params) {

        params.putAll(extraParamsMap);

        params.putAll(encodeParams(params));

        return this.getRetrofit()
                .getSpaceService()
                .getSpace(this.getPubnub().getConfiguration().getSubscribeKey(), spaceId, params);
    }

    @Override
    protected PNGetSpaceResult createResponse(Response<EntityEnvelope<PNSpace>> input) throws PubNubException {
        PNGetSpaceResult.PNGetSpaceResultBuilder resultBuilder = PNGetSpaceResult.builder();

        if (input.body() != null) {
            resultBuilder.space(input.body().getData());
        }
        return resultBuilder.build();
    }

    @Override
    protected PNOperationType getOperationType() {
        return PNOperationType.PNGetSpaceOperation;
    }

    @Override
    protected boolean isAuthRequired() {
        return true;
    }

    @Override
    public GetSpace includeFields(PNSpaceFields... params) {
        return appendInclusionParams(extraParamsMap, params);
    }

    @Override
    public TokenManagerProperties getTmsProperties() {
        return TokenManagerProperties.builder()
                .pnResourceType(PNResourceType.SPACE)
                .resourceId(spaceId)
                .build();
    }
}

File: pubnub-entities/src/main/kotlin/com/pubnub/entities/objects/endpoint/channel/GetSpace.kt
package com.pubnub.entities.objects.endpoint.channel

import com.pubnub.api.PubNub
import com.pubnub.api.endpoints.Endpoint
import com.pubnub.entities.objects.endpoint.internal.ReturningCustom
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.managers.RetrofitManager
import com.pubnub.entities.objects.space.PNSpace
import com.pubnub.entities.objects.space.PNSpaceResult
import com.pubnub.api.models.server.objects_api.EntityEnvelope
import retrofit2.Call
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.create
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.QueryMap

/**
 * @see [PubNub.getSpace]
 */
class GetSpace internal constructor(
    pubnub: PubNub,
    private val channel: String,
    private val withCustom: ReturningCustom
) : Endpoint<EntityEnvelope<PNSpace>, PNSpaceResult>(
    pubnub,
    pubnub.telemetryManager,
    pubnub.retrofitManager,
    pubnub.tokenManager
) {
    override fun getAffectedChannels(): List<String?> = listOf(channel)

    override fun getAffectedChannelGroups(): List<String?> = listOf()

    override fun validateParams() {
    }

    override fun doWork(queryParams: MutableMap<String, String>): Call<EntityEnvelope<PNSpace>> {
        val params = queryParams + withCustom.createIncludeQueryParams()
        return pubnub.retrofitManager.objectsService().getChannelMetadata(
            subKey = pubnub.configuration.subscribeKey,
            channel = channel,
            options = params
        )
    }

    override fun getOperationType(): PNOperationType? = PNOperationType.PNGetChannelMembersOperation

    override fun isAuthRequired(): Boolean = true

    override fun createResponse(input: Response<EntityEnvelope<PNSpace>>): PNSpaceResult? {
        return input.body()?.let {
            PNSpaceResult(
                status = it.status,
                data = it.data
            )
        }
    }
}

interface ObjectsService {
    @GET("v2/objects/{subKey}/channels/{channel}")
    fun getChannelMetadata(
        @Path("subKey") subKey: String,
        @Path("channel") channel: String,
        @QueryMap(encoded = true) options: Map<String, String> = mapOf()
    ): Call<EntityEnvelope<PNSpace>>
}

object Lazy {
    private lateinit var retrofit: Retrofit
    private val objectsService: ObjectsService by lazy {
        retrofit.create()
    }

    fun objectsService(retrofit: Retrofit): ObjectsService {
        this.retrofit = retrofit
        return objectsService
    }
}

fun RetrofitManager.objectsService(): ObjectsService {
    return Lazy.objectsService(this.transactionInstance)
}
